### PR TITLE
ENH: Add orient=tight format for dictionaries

### LINKED ISSUE #4889: to_json of df with multi-index header, followed by read_json results in incorrectly formatted column headers
cc @Komnomnomnom

I'm using a recent anaconda build on Windows, which includes v 0.12.0 pandas.  I'm attempting to use a multi-index header, write it out to a json file, import it and get the same formatted dataframe.  The column headers come out stacked vertically instead.

import numpy as np
import pandas as pd
pd.set_option('html',False)

index = pd.MultiIndex(levels=[['foo', 'bar', 'baz', 'qux'],
                           ['one', 'two', 'three']],
                   labels=[[0, 0, 0, 1, 1, 2, 2, 3, 3, 3],
                           [0, 1, 2, 0, 1, 1, 2, 0, 1, 2]],
                   names=['foo', 'bar'])
df = pd.DataFrame(np.random.randn(10, 3), index=index,
                  columns=['A', 'B', 'C'])
dft = df.transpose()
dft
dft.to_json('D:\xxmt.json',orient='split')
dftj=pd.read_json('D:\xxmt.json',orient='split')
dftj

### PR DESCRIPTION
**Pull request by Dr-Irv**
- [x] xref #4889 (see below)
- [x] tests added / passed
  - tests.frame.methods.test_to_dict.test_to_dict_orient_tight()
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

The issue in #4889 contains examples of how the JSON format doesn't support `MultiIndex` for the index or the columns, independent of the `orient` chosen.  Also, none of the orientations support index names (or `MultiIndex` names).  Finally, if you want to have a JSON format that is "tight" (or "compact"), the `split` is the closest thing, but it is incomplete due to the indexing issues.

As a first step to addressing this, I have created a `orient='tight'`  option to `DataFrame.to_dict()` and `DataFrame.from_dict()` .  If we agree that this is a reasonable representation, the next step (which I'd prefer to do in a second PR) would be to also support it with JSON.  The challenge there is that `to_json()` is currently implemented in C.  Right now, one can just use `json.dumps(df.to_dict(orient='tight'))` to get the needed JSON as a workaround.

I'm open to changing the word `'tight'` to something else, but I wanted it to start with a different letter than the other orientations.

### CODE REVIEW

**Comment by WillAyd** [comment]
Hmm +/- 0 on adding anything to to_dict here; we already have a lot of formats that kind of do the same things but not really and it's not always spelled out. While this could have some utility I think it adds to that confusion

> The challenge there is that `to_json()` is currently implemented in C.

If the ultimate goal is to include in the table schema most of that is written in Python, so maybe not as much of a concern?

**Comment by Dr-Irv** [comment]
> Hmm +/- 0 on adding anything to to_dict here; we already have a lot of formats that kind of do the same things but not really and it's not always spelled out. While this could have some utility I think it adds to that confusion
> 
> > The challenge there is that `to_json()` is currently implemented in C.
> 
> If the ultimate goal is to include in the table schema most of that is written in Python, so maybe not as much of a concern?

@WillAyd Here are two important comments in the original issue that this PR addresses:

https://github.com/pandas-dev/pandas/issues/4889#issuecomment-614798694  states (and I agree with this observation) that the `table` format is not compact when it comes to JSON.  The size of the 'table' format is not as compact as possible when you need to push JSON around as the payload in a REST API.

https://github.com/pandas-dev/pandas/issues/4889#issuecomment-645323225  states the issue with not preserving the `index` correctly.

File: pandas/core/frame.py [hunk 1]
```diff
@@ -1297,13 +1319,27 @@ def from_dict(cls, data, orient="columns", dtype=None, columns=None) -> "DataFra
...
         else:  # pragma: no cover
             raise ValueError("only recognize index or columns for orient")
 
-        return cls(data, index=index, columns=columns, dtype=dtype)
+        if orient != "tight":
+            return cls(data, index=index, columns=columns, dtype=dtype)
+        else:
+            realdata = data["data"]
+
+            def create_index(indexlist, namelist):
```
Comment (jreback) [inline pandas/core/frame.py:1333]: I *think* ensure_index does this

**Review by jreback — CHANGES_REQUESTED** [review]
some comments, looks pretty good if you'd update the notes to 1.2

**Comment by Dr-Irv** [comment]
@jreback before I make the change you suggested, when we had the dev sprint on 9/4/20, we discussed this PR, and it was suggested that I add an argument such as `preserve_names` to `to_dict()` that would apply to the `"split"` format rather than introduce the `"tight"` option.

So what is your opinion about introducing `orient="tight"` versus `orient="split", preserve_names=False` (where the latter would be used to preserve compatibility with prior versions)?

**Comment by simonjayhawkins** [comment]
@Dr-Irv what's the status here?

**Comment by Dr-Irv** [comment]
> @Dr-Irv what's the status here?

I've been waiting for @jreback to respond to this comment:  https://github.com/pandas-dev/pandas/pull/35292#issuecomment-699072742

I would like to introduce `orient="tight"`, but I'm not sure there is agreement on that.  If we can agree, I'll move forward with it.

**Comment by jreback** [comment]
looks ok, can you merge master and will look

**Comment by Dr-Irv** [comment]
@jreback have merged with master and all checks passed.  Changed it from draft status.

**Review by jreback — CHANGES_REQUESTED** [review]
looks good a comment

**Review by jreback — APPROVED** [review]

**Pull request merged**